modules/tcp/scanner.go

```diff
@@ -0,0 +1,234 @@
+}
+
+// ScanResults instances are returned by the module's Scan function.
+type Results struct {
+	Response string `json:"response,omitempty"`
+}
+
+// NewFlags returns an empty Flags object.
+func (module *Module) NewFlags() interface{} {
+	return new(Flags)
+}
+
+// NewScanner returns a new instance Scanner instance.
+func (module *Module) NewScanner() zgrab2.Scanner {
+	return new(Scanner)
+}
+
+// Description returns an overview of this module.
+func (module *Module) Description() string {
+	return "Send a TCP probe and read the response"
+}
+
+// Validate performs any needed validation on the arguments
+func (flags *Flags) Validate(args []string) error {
+	if flags.Protocol != "\\n" && flags.ProbeFile != "" {
+		log.Fatal("Cannot set both --probe and --probe-file")
+		return zgrab2.ErrInvalidArguments
+	}
+	return nil
+}
+
+// Help returns module-specific help
+func (flags *Flags) Help() string {
+	return ""
+}
+
+// Protocol returns the protocol identifer for the scanner.
+func (scanner *Scanner) Protocol() string {
+	return "tcp"
+}
+
+// GetName returns the name defined in the Flags.
+func (scanner *Scanner) GetName() string {
+	return scanner.config.Name
+}
+
+// GetTrigger returns the Trigger defined in the Flags.
+func (scanner *Scanner) GetTrigger() string {
+	return scanner.config.Trigger
+}
+
+// InitPerSender does nothing in this module.
+func (scanner *Scanner) InitPerSender(senderID int) error {
+	return nil
+}
+
+// Init initializes the Scanner with the command-line flags.
+func (scanner *Scanner) Init(flags zgrab2.ScanFlags) error {
+	var err error
+	f, _ := flags.(*Flags)
+	scanner.config = f
+	if len(f.ProbeFile) != 0 {
+		scanner.probe, err = ioutil.ReadFile(f.ProbeFile)
+		if err != nil {
+			log.Fatal("Failed to open probe file")
+			return zgrab2.ErrInvalidArguments
+		}
+	} else {
+		if len(scanner.probe) == 0 {
+			strProtocol, err := strconv.Unquote(fmt.Sprintf(`"%s"`, scanner.config.Protocol))
+			strProbe := ""
+			opcode := ""
+			if err != nil {
+				panic("Probe error")
+			}
+			//determine which probe to use
+			if strProtocol == "pptp" {
+				strProbe = "\x00\x9c\x00\x01\x1a\x2b\x3c\x4d\x00\x01\x00\x00\x01\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x4d\x69\x63\x72\x6f\x73\x6f\x66\x74\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+				scanner.probe = []byte(strProbe)
+			}
+			if strProtocol == "openvpn" {
+				length := ""
+
+				//generate session ID
+				//the ID is static in order to identify and verify incoming responses
+				session := "1a2b3c4d1a2b3c4d"
```

Comment:
It makes sense to use a static session ID in order to identify and verify the responses, but it seems a little weird to do that when publishing the modules. Should we maybe give the user to set a custom session ID?

Reply:
You can implement this as a config parameter and have this session ID specified as the default there.